### PR TITLE
FDMPC: generate pyop2 kernels

### PR DESCRIPTION
**Pull request by pbrubeck**
# Description
Previously, FDMPC assembly was done entirely with petsc4py calls inside python loops across the cells.
This PR implements the custom FDM assembly with pyop2 kernels.

### CODE REVIEW

File: firedrake/preconditioners/pmg.py [hunk 1]
```diff
@@ -602,16 +602,18 @@ def get_readonly_view(arr):
 
 
 @cached({}, key=generate_key_evaluate_dual)
-def evaluate_dual(source, target, alpha=tuple()):
+def evaluate_dual(source, target, derivative=None):
```
Comment (connorjward) [inline firedrake/preconditioners/pmg.py:605]: Can this get a numpy-style docstring? At the very least what is written doesn't make it clear what the types are.

File: firedrake/preconditioners/fdm.py [hunk 2]
```diff
@@ -159,35 +161,36 @@ def initialize(self, pc):
...
-        self.pc = fdmpc
         if hasattr(self, "_ctx_ref"):
             with dmhooks.add_hooks(self._dm, self, appctx=self._ctx_ref, save=False):
                 fdmpc.setFromOptions()
         else:
             fdmpc.setFromOptions()
 
     @PETSc.Log.EventDecorator("FDMPrealloc")
-    def allocate_matrix(self, V, J, bcs, fcp, pmat_type, use_static_condensation):
+    def allocate_matrix(self, Amat, V, J, bcs, fcp, pmat_type, use_static_condensation, use_amat):
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:193]: Docstring now out-of-date.

File: firedrake/preconditioners/fdm.py [hunk 3]
```diff
@@ -197,8 +200,9 @@ def allocate_matrix(self, V, J, bcs, fcp, pmat_type, use_static_condensation):
         :arg fcp: form compiler parameters to assemble coefficients
         :arg pmat_type: the `PETSc.Mat.Type` for the blocks in the diagonal
         :arg use_static_condensation: are we assembling the statically-condensed Schur complement on facets?
+        :arg use_amat: are we computing the Schur complement exactly?
 
-        :returns: 2-tuple with the preconditioner :class:`PETSc.Mat` and a list of assembly callables
+        :returns: 3-tuple with the stiffness and preconditioner :class:`PETSc.Mats` and a list of assembly callables
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:205]: I expect that :class:`PETSc.Mats` won't resolve correctly in the RST, please can you check?

File: firedrake/preconditioners/fdm.py [hunk 4]
```diff
@@ -233,71 +245,83 @@ def allocate_matrix(self, V, J, bcs, fcp, pmat_type, use_static_condensation):
...
+            P.setType(ptype)
+            P.setSizes(sizes)
+            P.setPreallocationNNZ((dnz, onz))
+            P.setOption(PETSc.Mat.Option.IGNORE_OFF_PROC_ENTRIES, False)
+            P.setOption(PETSc.Mat.Option.NEW_NONZERO_ALLOCATION_ERR, True)
+            P.setOption(PETSc.Mat.Option.UNUSED_NONZERO_LOCATION_ERR, True)
+            P.setOption(PETSc.Mat.Option.STRUCTURALLY_SYMMETRIC, on_diag)
+            if ptype.endswith("sbaij"):
+                P.setOption(PETSc.Mat.Option.IGNORE_LOWER_TRIANGULAR, True)
+            # P.setLGMap(Vrow.dof_dset.lgmap, Vcol.dof_dset.lgmap)
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:298]: Old code should be deleted.

File: firedrake/preconditioners/fdm.py [hunk 5]
```diff
@@ -233,71 +245,83 @@ def allocate_matrix(self, V, J, bcs, fcp, pmat_type, use_static_condensation):
...
-                        assembly_callables.append(partial(P.setValuesRCV, bdofs, bdofs, vals, addv))
+                Pmats[Vrow, Vcol] = PETSc.Mat().createTranspose(Pmats[Vcol, Vrow])
+                continue
+
+            # Preallocate and assemble the FDM auxiliary sparse operator
+            on_diag = Vrow == Vcol
+            sizes = tuple(Vsub.dof_dset.layout_vec.getSizes() for Vsub in (Vrow, Vcol))
+            ptype = pmat_type if on_diag else PETSc.Mat.Type.AIJ
+
+            preallocator = PETSc.Mat().create(comm=self.comm)
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:276]: Can this not be achieved by using PyOP2 `Sparsity`? Why are you hand-rolling this matrix instantiation code?

File: firedrake/preconditioners/fdm.py [hunk 6]
```diff
@@ -343,8 +367,58 @@ def destroy(self, pc):
             self.pc.getOperators()[-1].destroy()
             self.pc.destroy()
 
+    def condense(self, A, J, bcs, fcp, pc_type="icc"):
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:370]: I think that this needs a docstring explaining what is going on.

File: firedrake/preconditioners/fdm.py [hunk 7]
```diff
@@ -569,219 +648,162 @@ def set_values(self, A, Vrow, Vcol, addv, triu=False):
         :arg Vrow: the :class:`.FunctionSpace` test space
         :arg Vcol: the :class:`.FunctionSpace` trial space
         :arg addv: a `PETSc.Mat.InsertMode`
-        :arg triu: are we assembling only the upper triangular part?
+        :arg mat_type: a `PETSc.Mat.Type`
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:651]: Please make this more descriptive. What is it used for?

File: firedrake/preconditioners/fdm.py [hunk 8]
```diff
@@ -569,219 +648,162 @@ def set_values(self, A, Vrow, Vcol, addv, triu=False):
...
-                insert(A, kernel(*self.kernel_args, result=result),
-                       self.map_rows(), self.map_cols(), addv)
-        else:
-            for e in range(self.nel):
-                self.set_indices(e)
-                for _ in range(self.layers[e]):
-                    insert(A, kernel(*self.kernel_args, result=result),
-                           self.map_rows(), self.map_cols(), addv)
-                    self.add_offsets()
+        if A.getType() == "preallocator":
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:684]: Perhaps this block of code could do with a comment.

File: firedrake/preconditioners/fdm.py [hunk 9]
```diff
@@ -569,219 +648,162 @@ def set_values(self, A, Vrow, Vcol, addv, triu=False):
...
-        return result or self.result
-
-    def __del__(self):
-        self.destroy()
+    code = """
+PetscErrorCode %(name)s(const Mat A, const Mat B, %(indices)s) {
+    PetscFunctionBeginUser;
+    PetscCall(MatSetValuesSparse(A, B, %(rows)s, %(cols)s, %(addv)d));
+    PetscFunctionReturn(PETSC_SUCCESS);
+}"""
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:700]:
```suggestion
    code = dedent("""
        PetscErrorCode %(name)s(const Mat A, const Mat B, %(indices)s) {
            PetscFunctionBeginUser;
            PetscCall(MatSetValuesSparse(A, B, %(rows)s, %(cols)s, %(addv)d));
            PetscFunctionReturn(PETSC_SUCCESS);
        }""")
```

https://docs.python.org/3/library/textwrap.html#textwrap.dedent

Just for readability. Could be applied to other C-strings as well.

File: firedrake/preconditioners/fdm.py [hunk 10]
```diff
@@ -569,219 +648,162 @@ def set_values(self, A, Vrow, Vcol, addv, triu=False):
...
 
-    def destroy(self):
-        pass
+    def __init__(self, A, name=None):
+        self.result = A
+        self.mats = [self.result]
+        self.name = name or type(self).__name__
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:705]: This doesn't seem like a very unique name. Is that OK?

File: firedrake/preconditioners/fdm.py [hunk 11]
```diff
@@ -569,219 +648,162 @@ def set_values(self, A, Vrow, Vcol, addv, triu=False):
...
+        self.rules = {}
+
+    def make_args(self, *mats):
+        return [op2.PassthroughArg(op2.OpaqueType(mat.klass), mat.handle) for mat in list(mats) + self.mats]
+
+    def kernel(self, mat_type="aij", on_diag=False, addv=None):
+        if addv is None:
+            addv = PETSc.InsertMode.INSERT
+        indices = ("rindices", "cindices")[:2-on_diag]
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:714]:
```suggestion
        indices = ("rindices",) if on_diag else ("rindices", "cindices")
```
? I think is more readable.

File: firedrake/preconditioners/fdm.py [hunk 12]
```diff
@@ -569,219 +648,162 @@ def set_values(self, A, Vrow, Vcol, addv, triu=False):
...
+        code = ""
+        if "MatSetValuesArray" in self.code:
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:716]: This seems like an odd thing to check.

File: firedrake/preconditioners/fdm.py [hunk 13]
```diff
@@ -852,15 +874,48 @@ def condense(self, result=None):
 
 
 class SchurComplementBlockCholesky(SchurComplementKernel):
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:876]: I feel like all of these classes should have proper docstrings explaining what they do. I at least do not have the mathematical background to know what they do just from the names.

File: firedrake/preconditioners/fdm.py [hunk 14]
```diff
@@ -982,75 +1110,234 @@ def condense(self, result=None):
...
-                      PetscInt *cindices,
-                      InsertMode addv)
-{{
-    PetscInt ncols, irow, icol;
-    PetscInt *cols, *indices;
-    PetscScalar *vals;
-
-    PetscInt m, n;
-    PetscErrorCode ierr;
+def wrap_form(a, prefix="form", fcp=None, matshell=False):
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:1113]: Please provide a docstring.

**Review by connorjward — CHANGES_REQUESTED** [review]
Thanks for the work. I would like this to have a lot more documentation than it currently has.

File: firedrake/preconditioners/fdm.py (hunk 5 above)
Comment (pbrubeck) [reply firedrake/preconditioners/fdm.py:276]: Does that allow sparse element matrices or just dense blocks?

Comment (connorjward) [reply firedrake/preconditioners/fdm.py:278]: No. I now see what's going on.

File: firedrake/preconditioners/fdm.py (hunk 10 above)
Comment (pbrubeck) [reply firedrake/preconditioners/fdm.py:737]: This is not relay an issue, as the generated code is disk cached with a has key.

File: firedrake/preconditioners/fdm.py (hunk 12 above)
Comment (pbrubeck) [reply firedrake/preconditioners/fdm.py:748]: Only if `MatSetValuesArray` is used then I define it.

File: firedrake/preconditioners/fdm.py [hunk 15]
```diff
@@ -343,8 +369,84 @@ def destroy(self, pc):
...
+        KSP object that is reused on each cell within an MPI rank.
+
+        Parameters
+        ----------
+        A :
+            The :class:`PETSc.Mat` to statically condense.
+        J :
+            The bilinear :class:`ufl.Form` to statically condense.
+        bcs :
+            An iteratable of boundary conditions.
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:384]:
```suggestion
            An iterable of boundary conditions.
```

File: firedrake/preconditioners/fdm.py [hunk 16]
```diff
@@ -343,8 +369,84 @@ def destroy(self, pc):
...
 
+    def condense(self, A, J, bcs, fcp, pc_type="icc"):
+        """Construct block matrices used for matrix-free static condensation.
+        The inversion of the interior-interior block is replaced with a local
+        KSP object that is reused on each cell within an MPI rank.
+
+        Parameters
+        ----------
+        A :
+            The :class:`PETSc.Mat` to statically condense.
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:380]:
```suggestion
        A : PETSc.Mat
            The matrix to statically condense.
```

File: firedrake/preconditioners/fdm.py [hunk 17]
```diff
@@ -560,228 +667,177 @@ def _element_mass_matrix(self):
         return PETSc.Mat().createAIJ((nrows, nrows), csr=(ai, aj, data), comm=PETSc.COMM_SELF)
 
     @PETSc.Log.EventDecorator("FDMSetValues")
-    def set_values(self, A, Vrow, Vcol, addv, triu=False):
+    def set_values(self, A, Vrow, Vcol, addv, mat_type="aij"):
         """
-        Assemble the stiffness matrix in the FDM basis using sparse reference
+        Assemble the auxiliary operator in the FDM basis using sparse reference
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:672]: numpy docstring please

File: firedrake/preconditioners/fdm.py [hunk 18]
```diff
@@ -982,75 +1152,266 @@ def condense(self, result=None):
...
+        V = u.function_space()
+        F = a(v, ufl.Coefficient(V))
+        kernels = compile_form(F, prefix, parameters=fcp)
+        kernel = kernels[-1].kinfo.kernel
+        nargs = len(kernel.arguments) - len(a.arguments())
+        ncoef = nargs - len(extract_firedrake_constants(F))
+        ctx_struct = "".join("const PetscScalar *restrict c%d, " % i for i in range(nargs))
+        ctx_pack = "const PetscScalar *appctx[%d] = {%s};" % (nargs, ", ".join("c%d" % i for i in range(nargs)))
+        ctx_coefficients = "".join("appctx[%d], " % i for i in range(ncoef))
+        ctx_constants = "".join(", appctx[%d]" % i for i in range(ncoef, nargs))
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:1199]: Use f-strings please.

**Review by connorjward — CHANGES_REQUESTED** [review]
only some small comments, this looks pretty good

File: firedrake/preconditioners/fdm.py [hunk 19]
```diff
@@ -982,75 +1152,266 @@ def condense(self, result=None):
...
+    Parameters
+    ----------
+    a : :class:`ufl.Form`
+        The bilinear form.
+    prefix : `str`
+        The kernel prefix.
+    fcp : `dict`
+        The form compiler parameters.
+    matshell : `bool`
+        A flag to wrap the kernel with a :class:`PETSc.Mat` of type shell.
```
Comment (pbrubeck) [inline firedrake/preconditioners/fdm.py:1167]:
```suggestion
    prefix : str
        The kernel prefix.
    fcp : dict
        The form compiler parameters.
    matshell : bool
        A flag to wrap the kernel with a :class:`PETSc.Mat` of type shell.
```

File: firedrake/preconditioners/fdm.py [hunk 20]
```diff
@@ -1304,6 +1670,86 @@ def get_base_elements(e):
     return [e]
 
 
+class SparseAssembler(object):
+
+    _cache = {}
```
[the remark this answers is not in the document]
Comment (pbrubeck) [reply firedrake/preconditioners/fdm.py:1682]: This is just caching the returned value of `pyop2.compilation.load(..., comm=comm)`

**Review by dham — APPROVED** [review]

**Review by connorjward — APPROVED** [review]

**Pull request merged**